docs/spec_draft.md

```diff
@@ -358,6 +359,56 @@ IEEE-754 specification.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.concatenate
+
+### Semantics
+
+Concatenates sequence of tensors in `val` along dimension `dimension` and
+produces a tensor as a result. The tensors are of the same rank as the input
+operands except the specified dimension.
```

Comment:
Similarly to the feedback that I left for #172, this section doesn't really explain how to obtain the `result` tensor from the `val` tensor. "Concatenates sequence of tensors in `val` along dimension `dimension`" sounds fancy but is not actionable.

Let's define op semantics from first principles, using only the definitions in this spec (see the "Types" section in  [spec_draft.md](https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#types)) and college maths.

Reply:
I left the more formal notation out of the spec since calculating the actual indices of the `result` tensor is very ugly. Here are some proposals for how to approach this:
1. Keep the semantics the same as [XLA doc](https://www.tensorflow.org/xla/operation_semantics#concatenate) (aka not specifying the details).
2. Write the details as:
![Screen Shot 2022-10-03 at 5 28 18 PM](https://user-images.githubusercontent.com/31293323/193709133-a9ef9d96-48cc-4f06-94f8-3c25cc8a590d.png)
which is a simplified version of:
![Screen Shot 2022-10-03 at 5 29 37 PM](https://user-images.githubusercontent.com/31293323/193709237-9df9865b-120e-4f99-88a3-8e94e9f230b5.png)
<!-- $$result = \left(\left(inputs[t][dimension] * Right\right)(i + 1)\right)\Biggr|_{t=0}^{n-1}\Biggr|_{i}^{Left}$$ -->
where:
$$Left=\prod_{k=0}^{dimension-1}inputs[k]$$
$$Right=\prod_{l=dimension+1}^{R-1}inputs[l]$$
<!-- $$\left(i * inputs[t][dimension] * Right + Right * inputs[t][dimension]\right)\Biggr|_{t=0}^{n-1}\Biggr|_{i}^{Left}$$ -->
2.5 The lexicographic ordering of the indices of each tensor `t_i` is calculated such that it takes `x` elements from each tensor `t_i` until all elements are used, where `x` is calculated as $$\prod_{l=dimension}^{R-1}inputs[l]$$

An example of the above:
```
inputs = {
[ 0,  1,  2,  3,
  7,  8,  9, 10,
 14, 15, 16, 17,
 21, 22, 23, 24], // 2x2x4
[ 4, 11, 18, 25], // 2x2x1
[ 5,  6, 12, 13,
 19, 20, 26, 27] //2x2x2
}
dimension=2
```
`Right` is the product of the shapes "right of" the specified dimension plus the size of `dimension`, in this case just 4, 1, and 2 respectively for each `t` in `input` ("Left of" `dimension=2` would be `2x2=4` for all `t` in `input`). We take 4 elements from `t0`, 1 from `t1`, 2 from `t2` until all elements are taken, which result in:
```
result = [[ 0,  1,  2,  3,  4,  5,  6,
            7,  8,  9, 10, 11, 12, 13],
          [14, 15, 16, 17, 18, 19, 20,
           21, 22, 23, 24, 25, 26, 27]]
```
3. If you have a much simpler way to express this, please let me know!

Reply:
May be something along this line:

Given,
`vals` = {V0, V1, ..., Vn}
$d_{i}^{j}$ = dimension size of jth dimension in V[j] 

With the concatenation of `vals` along dimension `d`, the value of `result` at index `i0, i1,...,  id, ..., iR-1` is given by
```
result[i0, i1,...,  id, ..., iR-1] = V[k][i0, i1,...,  id', ..., iR-1]
```
such that
$id <  \sum(i=1, n, d_{i}^{d})$ and 
there exist a `k` such that $id > \sum(i=1, k-1, d_{i}^{d}) - 1$

and id' is defined as $id - \sum(i=1, k-1, d_{i}^{d})$

Reply:
@ghpvnist @sdasgup3 Thank you for the analysis of how the more formal spec for `concatenate` might look like! I agree that formulas that involve the canonical representation are pretty much inscrutable, and I like the idea of using multidimensional indices instead. How about we simplify that a bit further along the following lines:

```
Concatenates sequence of tensors in `inputs` along dimension `dimension` and
produces a tensor as a result. More formally, 
`result[i0, i1, ..., id, ..., iR-1] = inputs[k][i0, i1, ..., kd, ..., iR-1]`,
where:
  1. `id = d0 + d1 + ... + dk-1 + kd`.
  1. `d` is equal to `dimension`, and `d0`, `d1`, ... are `d`th dimension 
     sizes of `inputs`.
```

Reply:
Thanks @burmako for simplifying so much.

Reply:
Thanks both @burmako @sdasgup3 for your input!